src/ruby/puppetserver-lib/puppet/server/catalog.rb

```diff
@@ -0,0 +1,121 @@
+require 'puppet/server'
+
+module Puppet::Server::Catalog
+
+  def self.convert_java_args_to_ruby(hash)
+    Hash[hash.collect do |key, value|
+        # Stolen and modified from params_to_ruby in handler.rb
+        newkey = key.to_s
+        # Java::ClojureLang::Keywords retain their leading colon when
+        # converted to a string
+        newkey.slice!(0)
+
+        if value.java_kind_of?(Java::ClojureLang::IPersistentMap)
+          [newkey, convert_java_args_to_ruby(value)]
+        else
+          [newkey, value.java_kind_of?(Java::JavaUtil::List) ? value.to_a : value]
```

Comment:
This whole conditional feels to me more like a 3 branch switch on value's type. Would that be more readable or am totally bikeshedding? (I think Ruby's case wouldn't work well here either, so it'd end up being a 3 branch if/elsif/else...)